### PR TITLE
Add slot support for function apps

### LINKED ISSUE #6581: Add slot support for all az functionapp commands
Add slot support for all az functionapp commands

### PR DESCRIPTION
**Pull request by ankitkumarr**
@panchagnula, would love a review from you.

Fixes #6581, #9126 

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 1]
```diff
@@ -1016,46 +1016,74 @@ def create_webapp_slot(cmd, resource_group_name, webapp, slot, configuration_sou
...
-                                       None if clone_from_prod else configuration_source)
-        _generic_site_operation(cmd.cli_ctx, resource_group_name, webapp,
-                                'update_configuration', slot, site_config)
+        update_slot_configuration_from_source(cmd, client, resource_group_name, webapp, slot, configuration_source)
+
+    result.name = result.name.split('/')[-1]
+    return result
+
+
+def create_functionapp_slot(cmd, resource_group_name, name, slot, configuration_source=None):
```
[the remark this answers is not in the document]
Comment (ankitkumarr) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Slot creation seemed to work for Linux. Additionally, any such failure would ideally come from the backend APIs directly. :)

[the remark this answers is not in the document]
Comment (ankitkumarr) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Looks like slots work for dedicated linux. For consumption, there is some support (Linux Consumption is in preview), and work is planned for full support. So, we shouldn't block users on the CLI. This is the same behavior portal follows right now. Let me know if you have additional concerns.

**Comment by ankitkumarr** [comment]
@panchagnula, I have updated the PR to add an error message when use webapp's create command. Please have another look if you don't mind :)

**Comment by bigdatamoore** [comment]
Hi @ankitkumarr thanks for implementing this.  I was able to check out the repo and test this out and was able to create a slot.  I was wondering about the ability to deploy using az functionapp to slots.  I believe az webapp has a deployment option (az webapp deployment source config-zip) that has a --slot option.   Are you planning on adding the same for the az functionapp deployment commands?

https://docs.microsoft.com/en-us/cli/azure/webapp/deployment/source?view=azure-cli-latest#az-webapp-deployment-source-config-zip

**Comment by ankitkumarr** [comment]
@bigdatamoore, this PR should have addressed that as well. `webapp` and `functionapp` use a lot of the same functionality underneath. I turned slots on for all the `functionapp` commands that should have it. Let me know if you are unable to use it. Thanks!

**Comment by bigdatamoore** [comment]
@ankitkumarr  I was able to deploy a c# function using the az functionapp deployment source config-zip --slot command.  I tried it with a python function and got the following error:

```
(venv) PS > az functionapp deployment source config-zip --resource-group myresourcegroup --name myfunction1 --src test-function.zip --slot staging
Getting scm site credentials for zip deployment
Starting zip deployment. This operation can take a while to complete ...
The command failed with an unexpected error. Here is the traceback:

Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "C:\Users\240721\source\repos\azure-cli\venv\lib\site-packages\knack\cli.py", line 206, in invoke
    cmd_result = self.invocation.execute(args)
  File "c:\users\240721\source\repos\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 561, in execute
    raise ex
  File "c:\users\240721\source\repos\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 619, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "c:\users\240721\source\repos\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 612, in _run_job
    six.reraise(*sys.exc_info())
  File "C:\Users\240721\source\repos\azure-cli\venv\lib\site-packages\six.py", line 693, in reraise
    raise value
  File "c:\users\240721\source\repos\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 589, in _run_job
    result = cmd_copy(params)
  File "c:\users\240721\source\repos\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 298, in __call__
    return self.handler(*args, **kwargs)
  File "c:\users\240721\source\repos\azure-cli\src\azure-cli-core\azure\cli\core\__init__.py", line 467, in default_command_handler
    return op(**command_args)
  File "c:\users\240721\source\repos\azure-cli\src\command_modules\azure-cli-appservice\azure\cli\command_modules\appservice\custom.py", line 296, in enable_zip_deploy
    authorization, timeout)
  File "c:\users\240721\source\repos\azure-cli\src\command_modules\azure-cli-appservice\azure\cli\command_modules\appservice\custom.py", line 2233, in _check_zip_deployment_status
    res_dict = response.json()
  File "C:\Users\240721\source\repos\azure-cli\venv\lib\site-packages\requests\models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
  File "C:\Python36\lib\json\__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "C:\Python36\lib\json\decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Python36\lib\json\decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by ankitkumarr** [comment]
@bigdatamoore, oh are you on a Linux Consumption plan? Slots aren't yet supported for Linux Consumption, neither is the `az functionapp deployment source config-zip` command yet. I believe we are tracking support of these scenarios.

**Comment by bigdatamoore** [comment]
Ok thanks.  That is what I figured.  Thanks for the update.  I will wait until those are supported. Any idea as to when that could be?

**Comment by bigdatamoore** [comment]
@ankitkumarr I am seeing something weird with the c# function (http trigger).  I tried deploying a new version using config-zip --slot command to a staging slot.  However it appears to be deploying to both the production slot and the staging slot, as both urls are returning the same thing.  I could be doing something wrong.

Steps I followed:
1. Create a http trigger c# function
2. deploy the function to a new functionapp
3. Create a slot using az functionapp deployment slot create 
4. publish the function app, and create a zip file
5. deploy the function to the functionapp production slot by using az functionapp deployment source config-zip 
6. make a change to the app, publish  a new version and create a new zip file
7. deploy the modified function to the functionapp app staging slot by using az functionapp deployment source config-zip with the --slot staging option.
You will see that both the production slot and the staging slot have the same code

**Comment by bigdatamoore** [comment]
Hi @ankitkumarr, I took a look at the source code and although I have pretty limited understanding of how this works, I did find a few things that I thought might be causing what I am seeing.  Please correct me if I am off track.  

This appears to be where the zip deployment is happening.  I see there is a slot argument which if passed in, will be used to deploy to a slot.
https://github.com/Azure/azure-cli/blob/cbb0c773d681f5f0a779d145793fbfa8d45e7eca/src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py#L276

However when I searched for where this function is called from, I only found this reference:
https://github.com/Azure/azure-cli/blob/cbb0c773d681f5f0a779d145793fbfa8d45e7eca/src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py#L2479

I don't see it passing in the slot argument.

**Comment by bigdatamoore** [comment]
@ankitkumarr 
Ok nevermind.  I was able to get it working.   Sorry for the false alarm.   The problem I had was that in my staging slot I had the WEBSITE_CONTENTSHARE set to the same value as my production slot.  So when I updated one, it was automatically updating the other as well.

I have confirmed that the CLI is working with slots for C# functions.  I was able to create a slot, deploy to the slot, and then use the CLI to swap the slots.

**Comment by ankitkumarr** [comment]
@bigdatamoore, thanks for reporting this. I will take a look at your issue later today.

As for the actual code, you did find the correct function `enable_zip_deploy`. But, it is linked directly to the command here - https://github.com/Azure/azure-cli/blob/cbb0c773d681f5f0a779d145793fbfa8d45e7eca/src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/commands.py#L253. That's why you don't see any code references to it. 
The reference you linked is actually for a different command `az webapp up` that uses the function to deploy.

Anyway, I will try and get a repro to see what's going on.

**Comment by ankitkumarr** [comment]
@bigdatamoore, whoops, I missed you comment. Thanks for verifying. Glad you were able to resolve.

**Comment by bigdatamoore** [comment]
@ankitkumarr, can you test out the create slot command and see if it updates the WEBSITE_CONTENTSHARE setting value?  When doing this via the portal, the WEBSITE_CONTENTSHARE is automatically appended with a random string.  However when doing this via the CLI it is copying the same value that is in the production slot.  So that was actually causing the behavior that I was seeing.    Probably you will want to add some random string to the end of the existing value so that you don't have these collisions when using the CLI to create slots.  This may be by design but I am not sure.  Thanks.

**Pull request merged**